4.47.1+ started breaking tox-ansible pipelines

## Issue

4.47.1+ started breaking tox-ansible pipelines:

- [https://github.com/ansible/ansible-content-actions/blob/main/.github/workflows/sanity.yaml](https://github.com/ansible/ansible-content-actions/blob/main/.github/workflows/sanity.yaml)
- [https://github.com/ansible/ansible-content-actions/blob/main/.github/workflows/unit.yaml](https://github.com/ansible/ansible-content-actions/blob/main/.github/workflows/unit.yaml)

## Environment

Provide at least:

- OS: GitHub Actions

## Output of running tox

<details open>
<summary>Output of Pipeline</summary>

```console
Run python -m tox --ansible -e unit-py3.10-2.16 --conf tox-ansible.ini
  python -m tox --ansible -e unit-py3.10-2.16 --conf tox-ansible.ini
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.10.19/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.10.19/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.19/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.19/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.19/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.10.19/x64/lib
    GITHUB_TOKEN: ***
unit-py3.10-2.16: internal error
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/session/cmd/run/single.py", line 52, in _evaluate
    tox_env.setup()
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/tox_env/api.py", line 299, in setup
    self._setup_env()
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/tox_env/python/runner.py", line 145, in _setup_env
    super()._setup_env()
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/tox_env/python/api.py", line 286, in _setup_env
    self.ensure_python_env()
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/tox_env/python/api.py", line 290, in ensure_python_env
    conf = self.python_cache()
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/tox_env/python/virtual_env/api.py", line 94, in python_cache
    base = super().python_cache()
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/tox_env/python/api.py", line 327, in python_cache
    "version_info": list(self.base_python.version_info),
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/tox_env/python/api.py", line 333, in base_python
    base_pythons: list[str] = self.conf["base_python"]
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/config/sets.py", line 150, in __getitem__
    return self.load(item)
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/config/sets.py", line 162, in load
    return config_definition.__call__(self._conf, self.loaders, ConfigLoadArgs(chain, self.name, self.env_name))  # noqa: PLC2801
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/config/of_type.py", line 121, in __call__
    value = self.default(conf, args.env_name) if callable(self.default) else self.default  # ty: ignore[call-top-callable]
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/tox_env/python/api.py", line 177, in _base_python_default
    base_python = None if env_name is None else self.extract_base_python(env_name)
  File "/opt/hostedtoolcache/Python/3.10.19/x64/lib/python3.10/site-packages/tox/tox_env/python/api.py", line 198, in extract_base_python
    raise ValueError(msg)
ValueError: conflicting factors py3.10, 2.16 in unit-py3.10-2.16
  unit-py3.10-2.16: FAIL code 2 (0.00 seconds)
  evaluation failed :( (1.40 seconds)
Error: Process completed with exit code 2.
```

</details>

--- a/docs/changelog/3850.bugfix.rst
+++ b/docs/changelog/3850.bugfix.rst
@@ -1,0 +1,4 @@
+``ignore_base_python_conflict`` now also suppresses errors when an environment name contains multiple conflicting
+version-like factors (e.g., ``unit-py3.10-2.16`` where both ``py3.10`` and ``2.16`` are detected as Python versions).
+Previously this flag only handled conflicts between explicit ``base_python`` settings and the environment name - by
+:user:`gaborbernat`.

--- a/docs/reference/config.rst
+++ b/docs/reference/config.rst
@@ -521,7 +521,7 @@ Python language core options
 
 .. conf::
     :keys: ignore_base_python_conflict, ignore_basepython_conflict
-    :default: True
+    :default: False
     :version_added: 3.1
 
      .. versionadded:: 3.1.0
@@ -535,6 +535,17 @@ Python language core options
      Furthermore, we allow hard enforcing this rule by setting this flag to ``true``. In such cases we ignore the
      :ref:`base_python` and instead always use the base Python implied from the Python name. This allows you to configure
      :ref:`base_python` in the :ref:`base` section without affecting environments that have implied base Python versions.
+
+     **This flag handles two types of conflicts:**
+
+     1. **Explicit setting vs. environment name**: When :ref:`base_python` is explicitly set but conflicts with the
+        version extracted from the environment name. Setting this to ``true`` uses the version from the environment name.
+
+     2. **Multiple version factors in environment name** (since 4.47.3): When the environment name itself contains
+        multiple factors that look like Python versions (e.g., ``unit-py3.10-2.16`` where both ``py3.10`` and ``2.16``
+        match Python version patterns). Setting this to ``true`` falls back to :ref:`default_base_python` instead of
+        raising an error. To avoid this ambiguity, prefix non-Python version factors (e.g. use ``ac2.16`` instead of
+        ``2.16``).
 
 .. _conf-testenv:
 
@@ -1359,6 +1370,13 @@ Python options
     The preferred naming convention is ``N.M`` (e.g. ``3.14``, ``3.13``) rather than ``pyNMM`` (e.g. ``py314``,
     ``py313``). The dotted form is clearer, avoids ambiguity for Python versions >= 3.10, and reads more naturally in
     environment lists and CI output.
+
+    .. warning::
+
+       Any factor matching the pattern ``2.N`` or ``3.N`` (where N is one or more digits) is treated as a Python
+       version specifier. This means non-Python versions like ``2.16`` in an environment name such as
+       ``unit-py3.10-2.16`` will conflict with the ``py3.10`` factor. To avoid this, prefix non-Python version numbers
+       (e.g. ``ac2.16``), or set :ref:`ignore_base_python_conflict` to ``true``.
 
     .. versionadded:: 4.46
 

--- a/src/tox/tox_env/python/api.py
+++ b/src/tox/tox_env/python/api.py
@@ -174,7 +174,13 @@ class Python(ToxEnv, ABC):
         return env
 
     def _base_python_default(self, conf: Config, env_name: str | None) -> list[str]:  # noqa: ARG002
-        base_python = None if env_name is None else self.extract_base_python(env_name)
+        try:
+            base_python = None if env_name is None else self.extract_base_python(env_name)
+        except ValueError:
+            if self.core["ignore_base_python_conflict"]:
+                base_python = None
+            else:
+                raise
         return self.conf["default_base_python"] if base_python is None else [base_python]
 
     @classmethod
@@ -217,7 +223,12 @@ class Python(ToxEnv, ABC):
         base_pythons: list[str],
         ignore_base_python_conflict: bool,  # noqa: FBT001
     ) -> list[str]:
-        env_base_python = cls.extract_base_python(env_name)
+        try:
+            env_base_python = cls.extract_base_python(env_name)
+        except ValueError:
+            if ignore_base_python_conflict:
+                return base_pythons
+            raise
         if env_base_python is not None:
             spec_name = PythonSpec.from_string_spec(env_base_python)
             for base_python in base_pythons:

--- a/tests/tox_env/python/test_python_api.py
+++ b/tests/tox_env/python/test_python_api.py
@@ -55,6 +55,30 @@ def test_conflicting_base_python_factor_mixed_style() -> None:
         Python.extract_base_python("py310-3.11")
 
 
+def test_conflicting_base_python_factor_ignore(tox_project: ToxProjectCreator) -> None:
+    ini = "[tox]\nenv_list=unit-py3.10-2.16\nignore_base_python_conflict=true\n[testenv]\npackage=skip\n"
+    project = tox_project({"tox.ini": ini})
+    result = project.run("c", "-e", "unit-py3.10-2.16", "-k", "base_python")
+    result.assert_success()
+
+
+def test_conflicting_base_python_factor_no_ignore(tox_project: ToxProjectCreator) -> None:
+    ini = "[tox]\nenv_list=unit-py3.10-2.16\n[testenv]\npackage=skip\n"
+    project = tox_project({"tox.ini": ini})
+    result = project.run("c", "-e", "unit-py3.10-2.16", "-k", "base_python", raise_on_config_fail=False)
+    result.assert_failed(code=-1)
+
+
+def test_validate_base_python_conflicting_factors_ignore() -> None:
+    result = Python._validate_base_python("unit-py3.10-2.16", ["python3"], ignore_base_python_conflict=True)  # noqa: SLF001
+    assert result == ["python3"]
+
+
+def test_validate_base_python_conflicting_factors_no_ignore() -> None:
+    with pytest.raises(ValueError, match=r"conflicting factors py3\.10, 2\.16 in unit-py3\.10-2\.16"):
+        Python._validate_base_python("unit-py3.10-2.16", ["python3"], ignore_base_python_conflict=False)  # noqa: SLF001
+
+
 def test_build_wheel_in_non_base_pkg_env(
     tox_project: ToxProjectCreator,
     patch_prev_py: Callable[[bool], tuple[str, str]],

